In [ ]:
import pandas as pd 
import numpy as np




hello world


In [ ]:
from pathlib import Path
from statistics import NormalDist

import numpy as np
import pandas as pd

# Đọc dữ liệu (notebook nằm trong thư mục coding/, dữ liệu nằm trong dataset/)
duong_dan = Path("../dataset/sinh_vien.csv")
if not duong_dan.exists():
    duong_dan = Path("dataset/sinh_vien.csv")
df = pd.read_csv(duong_dan)

# Độ tin cậy 95% => alpha = 0.05
alpha = 0.05
# Phân vị chuẩn tắc z_{alpha/2}: giá trị z sao cho P(Z <= z) = 1 - alpha/2 = 0.975
z = NormalDist().inv_cdf(1 - alpha / 2)

print(f"Số quan sát: {len(df)}")
print(f"z_(alpha/2) = {z:.4f}")

## 3(c). Khoảng tin cậy 95% cho tỷ lệ tổng thể

**Biến chọn:** `lam_them` (nhị phân: Có / Không). Tham số cần ước lượng là $p$ – tỷ lệ sinh viên **có đi làm thêm** trong tổng thể.

**Công thức.** Gọi $n$ là cỡ mẫu, $m$ là số sinh viên có đi làm thêm trong mẫu, tỷ lệ mẫu là $f = \dfrac{m}{n}$ (ước lượng điểm của $p$). Khi $n$ lớn, theo định lý giới hạn trung tâm:

$$Z = \frac{f - p}{\sqrt{\dfrac{f(1-f)}{n}}} \;\approx\; N(0;1)$$

nên khoảng tin cậy $1-\alpha$ cho $p$ là

$$\left(f - z_{\alpha/2}\sqrt{\frac{f(1-f)}{n}}\;;\; f + z_{\alpha/2}\sqrt{\frac{f(1-f)}{n}}\right)$$

với $z_{\alpha/2} = z_{0{,}025} \approx 1{,}96$ là phân vị mức $1-\alpha/2$ của phân phối chuẩn tắc.

**Điều kiện áp dụng.**
- Mẫu ngẫu nhiên, các quan sát độc lập.
- Cỡ mẫu đủ lớn để xấp xỉ chuẩn: $nf \ge 5$ và $n(1-f) \ge 5$ (kiểm tra bằng code bên dưới).

In [ ]:
# --- 3(c): khoảng tin cậy cho tỷ lệ sinh viên có đi làm thêm ---
n = len(df)                           # cỡ mẫu
m = (df["lam_them"] == "Có").sum()    # số sinh viên có đi làm thêm
f = m / n                             # tỷ lệ mẫu (ước lượng điểm của p)

# Kiểm tra điều kiện xấp xỉ chuẩn
print(f"n = {n}, m = {m}, f = {f:.4f}")
print(f"n*f = {n * f:.1f} >= 5 ; n*(1-f) = {n * (1 - f):.1f} >= 5")

# Sai số chuẩn và độ chính xác (sai số ước lượng)
se_p = np.sqrt(f * (1 - f) / n)
eps_p = z * se_p

can_duoi_p, can_tren_p = f - eps_p, f + eps_p
print(f"Sai số chuẩn      = {se_p:.4f}")
print(f"Độ chính xác eps  = {eps_p:.4f}")
print(f"KTC 95% cho p     = ({can_duoi_p:.4f} ; {can_tren_p:.4f})")

n = 200, m = 79, f = 0.3950
n*f = 79.0 >= 5 ; n*(1-f) = 121.0 >= 5
Sai số chuẩn      = 0.0346
Độ chính xác eps  = 0.0678
KTC 95% cho p     = (0.3272 ; 0.4628)


**Nhận xét.** Trong mẫu có $79/200 = 39{,}5\%$ sinh viên đi làm thêm. Điều kiện xấp xỉ chuẩn thỏa mãn ($nf = 79$, $n(1-f) = 121$ đều lớn hơn 5).

Với độ tin cậy 95%, tỷ lệ sinh viên có đi làm thêm trong tổng thể nằm trong khoảng từ khoảng **32,7% đến 46,3%** (sai số ước lượng $\approx 6{,}8$ điểm phần trăm). Toàn bộ khoảng nằm dưới 50%, nên có thể nói sinh viên đi làm thêm là thiểu số. Khoảng này cũng chứa giá trị 0,4 – đúng bằng xác suất dùng khi mô phỏng dữ liệu.

*Ý nghĩa của "95%":* nếu lặp lại việc lấy mẫu nhiều lần và mỗi lần dựng một khoảng theo cách trên thì khoảng 95% số khoảng đó chứa giá trị $p$ thật.

## 3(d). Khoảng tin cậy 95% cho hiệu hai kỳ vọng $\mu_1 - \mu_2$

**Hai nhóm** (chia theo biến định tính `lam_them`):
- Nhóm 1: sinh viên **không** đi làm thêm,
- Nhóm 2: sinh viên **có** đi làm thêm.

**Biến định lượng so sánh:** `gio_tu_hoc_tuan` (số giờ tự học mỗi tuần). $\mu_1, \mu_2$ là số giờ tự học trung bình của hai nhóm trong tổng thể.

**Công thức.** Hai mẫu độc lập, phương sai tổng thể $\sigma_1^2, \sigma_2^2$ chưa biết, cỡ mẫu lớn ($n_1, n_2 > 30$). Thay $\sigma_i^2$ bằng phương sai mẫu hiệu chỉnh $s_i^2 = \dfrac{1}{n_i-1}\sum_{j}(x_{ij}-\bar{x}_i)^2$, ta có

$$Z = \frac{(\bar{x}_1 - \bar{x}_2) - (\mu_1 - \mu_2)}{\sqrt{\dfrac{s_1^2}{n_1} + \dfrac{s_2^2}{n_2}}} \;\approx\; N(0;1)$$

nên khoảng tin cậy $1-\alpha$ cho $\mu_1 - \mu_2$ là

$$\left(\bar{x}_1 - \bar{x}_2 \;\pm\; z_{\alpha/2}\sqrt{\frac{s_1^2}{n_1} + \frac{s_2^2}{n_2}}\right)$$

**Điều kiện áp dụng.**
- Hai mẫu ngẫu nhiên và độc lập với nhau (mỗi sinh viên chỉ thuộc một nhóm).
- $n_1 > 30$ và $n_2 > 30$: khi đó $\bar{x}_1 - \bar{x}_2$ xấp xỉ chuẩn theo định lý giới hạn trung tâm, không cần giả thiết tổng thể chuẩn và không cần giả thiết hai phương sai bằng nhau. (Biến `gio_tu_hoc_tuan` bị chặn dưới tại 1 nên không chuẩn tuyệt đối, nhưng cỡ mẫu lớn nên công thức vẫn dùng được.)

In [ ]:
# --- 3(d): khoảng tin cậy cho mu1 - mu2 (giờ tự học: không làm thêm - có làm thêm) ---
x1 = df.loc[df["lam_them"] == "Không", "gio_tu_hoc_tuan"].to_numpy()
x2 = df.loc[df["lam_them"] == "Có", "gio_tu_hoc_tuan"].to_numpy()

n1, n2 = len(x1), len(x2)

# Trung bình mẫu
tb1, tb2 = x1.sum() / n1, x2.sum() / n2

# Phương sai mẫu hiệu chỉnh (chia cho n - 1)
s1_bp = ((x1 - tb1) ** 2).sum() / (n1 - 1)
s2_bp = ((x2 - tb2) ** 2).sum() / (n2 - 1)

print(f"Nhóm 1 (không làm thêm): n1 = {n1}, trung bình = {tb1:.4f}, s1^2 = {s1_bp:.4f}")
print(f"Nhóm 2 (có làm thêm)   : n2 = {n2}, trung bình = {tb2:.4f}, s2^2 = {s2_bp:.4f}")

# Ước lượng điểm, sai số chuẩn và độ chính xác
hieu = tb1 - tb2
se_hieu = np.sqrt(s1_bp / n1 + s2_bp / n2)
eps_hieu = z * se_hieu

can_duoi_d, can_tren_d = hieu - eps_hieu, hieu + eps_hieu
print(f"Hiệu trung bình mẫu     = {hieu:.4f}")
print(f"Sai số chuẩn            = {se_hieu:.4f}")
print(f"Độ chính xác eps        = {eps_hieu:.4f}")
print(f"KTC 95% cho mu1 - mu2   = ({can_duoi_d:.4f} ; {can_tren_d:.4f})")

Nhóm 1 (không làm thêm): n1 = 121, trung bình = 16.1207, s1^2 = 30.7338
Nhóm 2 (có làm thêm)   : n2 = 79, trung bình = 13.5519, s2^2 = 22.1866
Hiệu trung bình mẫu     = 2.5688
Sai số chuẩn            = 0.7313
Độ chính xác eps        = 1.4334
KTC 95% cho mu1 - mu2   = (1.1354 ; 4.0021)


**Nhận xét.** Cả hai nhóm đều có cỡ mẫu lớn hơn 30 ($n_1 = 121$, $n_2 = 79$) nên điều kiện áp dụng thỏa mãn. Trong mẫu, sinh viên không đi làm thêm tự học trung bình 16,12 giờ/tuần, sinh viên có đi làm thêm tự học 13,55 giờ/tuần, chênh lệch khoảng 2,57 giờ.

Với độ tin cậy 95%, sinh viên không đi làm thêm tự học nhiều hơn sinh viên có đi làm thêm trung bình từ khoảng **1,14 đến 4,00 giờ mỗi tuần**. Khoảng tin cậy **không chứa 0** và nằm hoàn toàn bên phải 0, nên có cơ sở để kết luận $\mu_1 > \mu_2$: việc đi làm thêm đi kèm với số giờ tự học ít hơn. Khoảng này chứa giá trị 3 – đúng mức chênh lệch được cài vào khi mô phỏng dữ liệu.

Lưu ý: đây là dữ liệu mô phỏng nên kết luận chỉ mang ý nghĩa minh họa phương pháp; và khoảng tin cậy chỉ cho thấy có chênh lệch giữa hai nhóm, không tự nó chứng minh quan hệ nhân quả.